### PR TITLE
Add Feast subscription data to acquisition-event buffer part 1: Add Lambda and SQS cloudformation for apple and google handlers

### PR DESCRIPTION
**Pull request by michaelwmcnamara**
This work is the initial plumbing for two related pieces of work writing feast subscription data for android and iOS to the fact_acquisition_events table via the events buffer. 
Cards for this work: [android](https://trello.com/c/Og1NFpWC/270-feast-iap-acquisitions-should-appear-in-factacquisitionevent) and [iOS](https://trello.com/c/tTfjEXnv/379-feast-iap-acquisitions-should-appear-in-factacquisitionevent).

This change sets up the cloudformation, riffraff etc for this work but does not introduce any functionality.
This work is important because it provides the foundation for two related but independent pieces of work.
Merging this will allow the following work to proceed independently with no risk of overwriting because of changes to the same files

## What does this change?

Add new cloudformation file to handle these new lambda functions and related resources.
This is needed because the existing cloudformation file is at capacity and adding resources their results in cfn template failing validation due to size constraints.

Overall architecture of finished product:

image

Resources added in this PR:
- Feast-acquisition-events router Lambda function
- Feast-acquisition-events processing Lambda functions for apple and google devices
- SQS Queues to communicate between router Lambda function and processing lambdas, corresponding DLQ's and alerts
- IAM Profiles and Policies to enable desired functionality.

Update riffraff.yaml to include references to new cloudformation and lambda functions.
Update ci and webpack files with references to the new resources.

## How to test

- [x] Deploy to CODE
- [x] Ensure the new cloudformation and lambdas have been created.
- [x] Send a message from the apple SQS queue - confirm the apple lambda is called
- [x] Send a message from the google SQS queue - confirm the google lambda is called

##  Note about Deploying for the first time:
It is necessary to perform the following steps to deploy:
1: Preview the deploy, deselect all steps, then select only the steps that upload the lambda files to S3
2: Manually create the mobile-purchases-feast-write-acquisition-events-bus-{STAGE} stack using the feast-acquisition-events.cloudformation.yaml template - being sure to set the correct values for all parameters (Note most of these can be derived from the mobile-purchases stack).
3: Preview the deploy again, deselect all steps then select all steps associated with "mobile-purchases-feast-acquisition-events"

## How can we measure success?

New cloudformation, queues and lambdas have been created and work correctly. These can be used as a basis for further development. 

## Have we considered potential risks?

Minimal risk as no functionality is introduced with these lambdas beyond logging that they have been called.

### CODE REVIEW

File: typescript/src/feast/acquisition-events/apple.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    const message: String = 'Feast Apple Acquisition Events Lambda has been called';
+    console.log(message)
+    return message;
+    // const eventualEnsuredIdentityAccount = event.Records.flatMap(
+        // async (sqsRecord: SQSRecord) => {
+        //     console.log(
+        //         `Hello World`,
+        //     );
+        // });
+    // await Promise.all<void>(eventualEnsuredIdentityAccount);
```
Comment (tjmw) [inline typescript/src/feast/acquisition-events/apple.ts]: Is this needed or could it be removed?

File: typescript/src/feast/acquisition-events/google.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    const message: String = 'Feast Google Acquisition Events Lambda has been called';
+    console.log(message)
+    return message;
+    // const eventualEnsuredIdentityAccount = event.Records.flatMap(
+        // async (sqsRecord: SQSRecord) => {
+        //     console.log(
+        //         `Hello World`,
+        //     );
+        // });
+    // await Promise.all<void>(eventualEnsuredIdentityAccount);
```
Comment (rBangay) [inline typescript/src/feast/acquisition-events/google.ts:14]: Do you still need this comment in here?

File: typescript/tests/feast/acquisition-events/apple.test.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        {
+            "tickettailor-webhook-signature":
+            {
+                "stringValue": "t=1724160026,v1=a3dbd8cfb0f04a0a9b0dd9d2547f1dd1a51e60d528a4edaee3bc02085517bd50",
+                "stringListValues":
+                [],
+                "binaryListValues":
+                [],
+                "dataType": "String"
+            }
```
Comment (tjmw) [inline typescript/tests/feast/acquisition-events/apple.test.ts:26]: Could this be removed, to simplify the example payload?

File: .github/workflows/ci.yaml [hunk 4]
```diff
@@ -2,6 +2,10 @@ name: CI
 
 on: [push]
 
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
+
```
Comment (tjmw) [inline .github/workflows/ci.yaml:8]: Should this be pulled out to a separate PR? It feels quite distinct from the rest of the work in here. I think I'd vote to keep it separate, to make the different parts of the PR independently revertible if needed (for example).

File: typescript/src/feast/acquisition-events/google.ts (hunk 2 above)
Comment (michaelwmcnamara) [reply typescript/src/feast/acquisition-events/google.ts:14]: Its a placeholder I will be building on, but true I can probably cut it for the repo then paste it back in my own code :-)

File: riff-raff.yaml [hunk 5]
```diff
@@ -174,3 +168,31 @@ deployments:
     parameters:
       functionNames: [mobile-purchases-apple-revalidate-receipts-]
       fileName: apple-revalidate-receipts.zip
+
+  mobile-purchases-feast-acquisition-events-bus-cloudformation:
+    type: cloud-formation
+    parameters:
+      templatePath: feast-acquisition-events.cloudformation.yaml
+      appendStageToCloudFormationStackName: true
```
Comment (tjmw) [inline riff-raff.yaml:176]: I think `true` is the default, so we could remove this line and it would have no effect.

File: typescript/src/feast/acquisition-events/google.ts (hunk 2 above)
Comment (michaelwmcnamara) [reply typescript/src/feast/acquisition-events/google.ts]: removed now.

File: feast-acquisition-events.cloudformation.yaml [hunk 6]
```diff
@@ -0,0 +1,304 @@
+AWSTemplateFormatVersion: 2010-09-09
+Description: FeastGoogleAcquisitionEventsLambda
```
Comment (tjmw) [inline feast-acquisition-events.cloudformation.yaml:2]: I don't think this is Google specific now, right?

```suggestion
Description: FeastAcquisitionEventsLambda
```

File: typescript/src/feast/acquisition-events/apple.ts (hunk 1 above)
Comment (tjmw) [reply typescript/src/feast/acquisition-events/apple.ts]: Ha @rBangay got their first :-)

File: .github/workflows/ci.yaml (hunk 4 above)
Comment (michaelwmcnamara) [reply .github/workflows/ci.yaml:8]: thats interesting - I don't recognise that code. Perhaps its something that has changed in the main branch? I will try merging main and see if it disappears.

File: typescript/tests/feast/acquisition-events/apple.test.ts (hunk 3 above)
Comment (michaelwmcnamara) [reply typescript/tests/feast/acquisition-events/apple.test.ts]: Good point! Removed in latest commit.

File: .github/workflows/ci.yaml (hunk 4 above)
Comment (michaelwmcnamara) [reply .github/workflows/ci.yaml]: Nope seems to be something I added. I have removed it - I will confirm the code still builds and deploys.

File: riff-raff.yaml (hunk 5 above)
Comment (michaelwmcnamara) [reply riff-raff.yaml]: Done is latest commit

File: feast-acquisition-events.cloudformation.yaml (hunk 6 above)
Comment (michaelwmcnamara) [reply feast-acquisition-events.cloudformation.yaml]: Updated in latest commit.

File: feast-acquisition-events.cloudformation.yaml [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+          Value: !Ref Stage
+      VisibilityTimeout: 300
+    UpdateReplacePolicy: Delete
+    DeletionPolicy: Delete
+
+  FeastAppleAcquisitionEventsDlq:
+    Type: AWS::SQS::Queue
+    Properties:
+      QueueName: !Sub ${App}-${Stage}-feast-apple-acquisition-events-dlq
+      KmsMasterKeyId: alias/aws/sqs
```
Comment (johnduffell) [inline feast-acquisition-events.cloudformation.yaml:46]: what's the benefit of this over the default server side encryption that AWS provide? https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/creating-sqs-standard-queues.html#step-create-standard-queue:~:text=On%20August%2017%2C%202022%2C%20default%20server%2Dside%20encryption%20(SSE)%20was%20applied%20to%20all%20Amazon%20SQS%20queues.

File: feast-acquisition-events.cloudformation.yaml [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+        S3Key: !Sub ${Stack}/${Stage}/${App}-feast-apple-acquisition-events/feast-apple-acquisition-events.zip
+      Description: An SQS triggered lambda to process acquisition events for iOS devices and write to event buffer.
+      Environment:
+        Variables:
+          App: !Ref App
+          Stack: !Ref Stack
+          Stage: !Ref Stage
+          STACK: !Ref Stack
+          STAGE: !Ref Stage
+          APP: !Ref App
```
Comment (johnduffell) [inline feast-acquisition-events.cloudformation.yaml:200]: why both Lower and UPPER case?

File: feast-acquisition-events.cloudformation.yaml [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+                - - 'arn:aws:secretsmanager:'
+                  - !Ref AWS::Region
+                  - ':'
+                  - !Ref AWS::AccountId
+                  - !Sub :secret:${Stage}/${App}-feast-apple-acquisition-events/*
+                  - !Sub :secret:${Stage}/${App}-feast-google-acquisition-events/*
+        Version: '2012-10-17'
+      PolicyName: Secretmanageraccesspolicy79659139
+      Roles:
+        - !Ref FeastAcquisitionEventsLambdaServiceRole
```
Comment (johnduffell) [inline feast-acquisition-events.cloudformation.yaml:304]: any alarms needed? e.g. DLQ?
I'm not sure how "finished" this CFN is - something I find handy is to "self review" PRs like this to leave comments in anticipation of questions people might have.  Then people know which areas to look in more details and which are placeholders (and also things I had to look in more detail to get right).  This also helps github archaeologists in future!

File: typescript/tests/feast/acquisition-events/apple.test.ts [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+            "ApproximateFirstReceiveTimestamp": "1724077616692"
+        },
+        "messageAttributes": {},
+        "md5OfBody": "f76fca7a395b41f1dd0d9af3b1755ac1",
+        "eventSource": "aws:sqs",
+        "eventSourceARN": "arn:aws:sqs:eu-west-1:865473395570:ticket-tailor-webhook-queue-CODE",
+        "awsRegion": "eu-west-1"
+    };
+
+const sQSEvent: SQSEvent = {
```
Comment (johnduffell) [inline typescript/tests/feast/acquisition-events/apple.test.ts:23]: I'd prbabbyl go for `sqsEvent` rather than `sQSEvent`

File: .github/workflows/ci.yaml [hunk 11]
```diff
@@ -104,3 +104,9 @@ jobs:
               - tsc-target/feast-google-pubsub.zip
             mobile-purchases-feast-google-update-subscriptions:
               - tsc-target/feast-google-update-subscriptions.zip
+            mobile-purchases-feast-acquisition-events-bus-cloudformation:
+              - feast-acquisition-events.cloudformation.yaml
```
[the remark this answers is not in the document]
Comment (michaelwmcnamara) [reply .github/workflows/ci.yaml:108]: I initially thought the same and spent a few days trying to get this to work as a cdk configuration. It was only after that I noticed that the  mobile-purchases-feast-cloudformation is commented out in riffraff.yaml. I don't think cdk is actually working in this repo at the moment.

File: feast-acquisition-events.cloudformation.yaml (hunk 7 above)
Comment (michaelwmcnamara) [reply feast-acquisition-events.cloudformation.yaml:46]: Honestly I am just aping what is generated by cdk here. If you think it is better without it I can remove it.

File: feast-acquisition-events.cloudformation.yaml (hunk 8 above)
Comment (michaelwmcnamara) [reply feast-acquisition-events.cloudformation.yaml:200]: Good point - I will remove this.

File: feast-acquisition-events.cloudformation.yaml (hunk 9 above)
Comment (michaelwmcnamara) [reply feast-acquisition-events.cloudformation.yaml:304]: Added alert for Dead Letter Queues - will consider adding more alerts as the functionality evolves.

File: typescript/tests/feast/acquisition-events/apple.test.ts (hunk 10 above)
Comment (michaelwmcnamara) [reply typescript/tests/feast/acquisition-events/apple.test.ts:23]: Agreed. that is nicer.

File: feast-acquisition-events.cloudformation.yaml (hunk 7 above)
Comment (michaelwmcnamara) [reply feast-acquisition-events.cloudformation.yaml:62]: ...Having read through that documentation though, I'm thinking its likely that the existing cloudformation is just old. I'll strip our the encryption key in next commit and let the default encryption take over.

Comment (michaelwmcnamara) [reply feast-acquisition-events.cloudformation.yaml]: This is now done.

File: feast-acquisition-events.cloudformation.yaml (hunk 9 above)
Comment (michaelwmcnamara) [reply feast-acquisition-events.cloudformation.yaml]: These are now added.

File: feast-acquisition-events.cloudformation.yaml [hunk 12]
```diff
@@ -0,0 +1,362 @@
+AWSTemplateFormatVersion: 2010-09-09
```
Comment (michaelwmcnamara) [inline feast-acquisition-events.cloudformation.yaml:1]: These resources have been added in a separate cloudformation file for 2 reasons:

1) The current cdk infrastructure is not working (in fact the  dependencies for cdk-defined resources have been commented out in riffraff.yaml indicating the deploy process for these resources is currently inactive.

2) The existing cloudformation.cfn file is at capacity. When I tried to add resources to this file it failed validation due to being over the size restrictions. It can be seen that similar action has been taken for [Dynamo table resources](https://github.com/guardian/mobile-purchases/blob/main/dynamo.cloudformation.yaml) and for [Exports mobile purchase data](https://github.com/guardian/mobile-purchases/blob/main/exports-cloudformation.yaml).

File: feast-acquisition-events.cloudformation.yaml [hunk 13]
```diff
@@ -0,0 +1,486 @@
...
+    Description: The ARN of the SNS topic to send all the cloudwatch alarms to
+    Type: String
+
+Mappings:
+  StageVariables:
+    CODE:
+      Schedule: 'rate(365 days)'
+      AlarmActionsEnabled: FALSE
+    PROD:
+      Schedule: 'rate(30 minutes)'
```
Comment (tjmw) [inline feast-acquisition-events.cloudformation.yaml:30]: Is this needed? Is there a schedule element to any of this infrastructure?

**Review by tjmw — APPROVED** [review]
I think this looks good (with one minor comment), thanks for working through the feedback 👍

File: feast-acquisition-events.cloudformation.yaml (hunk 13 above)
Comment (michaelwmcnamara) [reply feast-acquisition-events.cloudformation.yaml]: No it is not. Well spotted! Removing in next commit.
